---
title: "EX.4"
author: "Loai Naom"
date: "25/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(gProfileR)
library(knitr)

library(GO.db)
library(compGenomRData)
library(DESeq2)
library(gage)
```

# Loading the Data
```{r}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))
#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))
#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
stringsAsFactors = TRUE)
#define the design formula
designFormula <- "~ group"
#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
colData = colData,
design = as.formula(designFormula))
dds <- DESeq(dds)
DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
```

# choosing the important genes with big LFC and significant P Value
```{r}

#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]
#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]
#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]
#get the list of genes of interest
genesOfInterest <- rownames(DE)
#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate')

```

# Question 1
```{r}
goResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate')

KeggResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'KEGG',
hier_filtering = 'moderate')

REACResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'REAC',
hier_filtering = 'moderate')

KeggResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'CORUM',
hier_filtering = 'moderate')

goResults <- goResults[order(goResults$precision),]
goResults <- goResults[order(goResults$p.value),]


```

```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]
#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]
# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))
#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
geneSet2 <- sample(rownames(normalizedCounts), 25)
geneSets <- list('top_GO_term' = geneSet1,
'random_set' = geneSet2)
# Using the defined gene sets, we’d like to do a group comparison between the case
# samples with respect to the control samples.
#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
ref = match(rownames(colData[colData$group =='CTRL',]),
colnames(normalizedCounts)),
samp = match(rownames(colData[colData$group == 'CASE',]),
colnames(normalizedCounts)),
gsets = geneSets, compare = 'as.group')
```

# Question 2
```{r}
gseaResults_paired <- gage(exprs = log2(normalizedCounts+1),
ref = match(rownames(colData[colData$group =='CTRL',]),
colnames(normalizedCounts)),
samp = match(rownames(colData[colData$group == 'CASE',]),
colnames(normalizedCounts)),
gsets = geneSets, compare = 'paired')

gseaResults_unpaired <- gage(exprs = log2(normalizedCounts+1),
ref = match(rownames(colData[colData$group =='CTRL',]),
colnames(normalizedCounts)),
samp = match(rownames(colData[colData$group == 'CASE',]),
colnames(normalizedCounts)),
gsets = geneSets, compare = 'unpaired')

gseaResults_1ongroup <- gage(exprs = log2(normalizedCounts+1),
ref = match(rownames(colData[colData$group =='CTRL',]),
colnames(normalizedCounts)),
samp = match(rownames(colData[colData$group == 'CASE',]),
colnames(normalizedCounts)),
gsets = geneSets, compare = '1ongroup')
```

# Question 3
```{r}
goResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate', significant = F)
plot(-log10(goResults$p.value), goResults$term.size)

# There is a positive correlation
```

# Question 4
```{r}
goResults <- gprofiler(query = genesOfInterest,
organism = 'hsapiens',
src_filter = 'GO',
hier_filtering = 'moderate')

goResults_top10 <- goResults[order(goResults$term.size),]

geneSet_10 <- unlist(strsplit(goResults_top10[1,]$intersection, ','))

normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
geneSet2 <- sample(rownames(normalizedCounts), 25)
geneSETS_10 <- list('top_GO_term' = geneSet_10,
'random_set' = geneSet2)

gseaResults_10 <- gage(exprs = log2(normalizedCounts+1),
ref = match(rownames(colData[colData$group =='CTRL',]),
colnames(normalizedCounts)),
samp = match(rownames(colData[colData$group == 'CASE',]),
colnames(normalizedCounts)),
gsets = geneSETS_10, compare = 'as.group')


```

# Question 5
```{r}
# clusterProfiler, GSEAplot, LIGER, gsEasy, fgsea, seqGSEA
```

# Question 6
```{r}
# the code in this question is based on a code from Biostar website. 

# the code worked when running it normally but didn't work with knit, I marked the code 
# chunk that made me problems with #. 


library(topGO)
library(org.Hs.eg.db)
library(dplyr)
library(data.table)

DEresults_df <- as.data.frame(DEresults)
rn <- rownames(DEresults)
DEresults_df$VALUE <- rn

rnp <- DEresults_df$padj

#create a named vector
names(rnp) <- genesOfInterest


is.vector(rnp, mode = "numeric")

selection <- function(DEresults){ return(DEresults$padj < 0.05)}
allGO2genes  <- annFUN.org(whichOnto="BP", feasibleGenes=NULL, mapping="org.Hs.eg.db", ID="SYMBOL")
GOdata <- new("topGOdata", description = "topGO session", ontology = "BP",
 allGenes = rnp , geneSel = selection,   GO2genes=allGO2genes,
 nodeSize = 10,
 annot = annFUN.GO2genes)


#results.ks <- runTest(GOdata, algorithm="classic", statistic="ks")
#goEnrichment <- GenTable(GOdata, KS=results.ks, orderBy="KS", topNodes=20)
#goEnrichment$KS <- as.numeric(goEnrichment$KS)
#goEnrichment <- goEnrichment[goEnrichment$KS<0.05,]
#goEnrichment <- goEnrichment[,c("GO.ID","Term","KS")]
#goEnrichment$Term <- gsub(" [a-z]*\\.\\.\\.$", "", goEnrichment$Term)
#goEnrichment$Term <- gsub("\\.\\.\\.$", "", goEnrichment$Term)
#goEnrichment$Term <- paste(goEnrichment$GO.ID, goEnrichment$Term, sep=", ")
#goEnrichment$Term <- factor(goEnrichment$Term, levels=rev(goEnrichment$Term))

# run the statistic tests

#resultFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")

#GenTable(GOdata, classicFisher = resultFisher, topNodes = 10)
#resultFisher

```


# Question 7
```{r}
library(biomaRt)

# The getLDS() (Get Linked Dataset) function provides functionality to link 2 BioMart data sets which each other and construct a query over the two data sets, which means retrieving homology data across species.

#Every analysis with biomaRt starts with selecting a BioMart database to use. The commands below will connect us to Ensembl’s most recent version of the Human Genes BioMart and for the C.elegans data. After inspection of what values and what attributes each mart object can get, the following is just an example of randomly chosen attributes and values :


human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl")

C.elegans <- useEnsembl("ensembl", dataset = "celegans_gene_ensembl")
  
comp_res <- getLDS(attributes = c("celegans_homolog_associated_gene_name","chromosome_name", "start_position"),
       filters = "with_celegans_homolog", values = T,
       mart = human,
       attributesL = c("ensembl_gene_id","chromosome_name","start_position"), 
       martL = C.elegans)

comp_res
```

# Question 8
```{r}
# This question didn't work for me, I tried to download so many data bases but non 
# of them worked :(

#library(msigdbr)
#library(clusterProfiler)
#library(GEOquery)

#entrez <- read.gmt('c2.all.v7.4.entrez.gmt')
#order(entrez)
#symbols <- read.gmt('c2.all.v7.4.symbols.gmt')

#msigdata <- read.table('geneset.txt')

#GEOset <- getGEO('GSE3446')

